with_cfg/private/setting.bzl

```diff
@@ -9,10 +9,11 @@ def validate_and_get_attr_name(setting):
         # buildifier: disable=no-effect
         setting.workspace_name
 
-        # Ensure that the hash, which is a (signed) 32-bit integer, is non-negative, so that it does
-        # not contain a dash, which is not allowed in attribute names. Also ensure that the
-        # attribute name starts with a letter as it needs to be a valid identifier.
-        return "s_{}_{}".format(hash(str(setting)) + 2147483648, setting.name)
+        # Ensure the attribute name starts with a letter and doesn't contain a dash as it needs
+        # to be a valid identifier. Achieves that by ensuring that the hash, which is a (signed)
+        # 32-bit integer, is non-negative, so that it doesn't contain a dash, and in addition,
+        # replacing any potential dashes in the setting name.
+        return "s_{}_{}".format(hash(str(setting)) + 2147483648, setting.name.replace("-", "_"))
```

Comment:
There could be other characters in target names that aren't allowed in attributes. 

How about using the function you wrote for the test to loop over the characters and replace everything that isn't alphanumeric with an underscore? That should deal with any issues of this kind.